From bba03a463b5545aea9bc7b454952ad1e7acb1fca Mon Sep 17 00:00:00 2001
From: Sanyog Kale <skale@habana.ai>
Date: Fri, 08 Apr 2022 12:53:28 +0300
Subject: [PATCH] [SW-77251] gaudi3: add onix device specific mem region api

Since onix has limited SRAM memory of 6 MB this patch add
changes to get onix specific SRAM memory address where
memory is re-used for LKD/FWUPD and GP/I2C communications.

Change-Id: I5a1af6f4220b58f74843be057725c8e79b139764
---

diff --git a/arch/gaudi3/arc/onix.c b/arch/gaudi3/arc/onix.c
index f51b029..987fd22 100644
--- a/arch/gaudi3/arc/onix.c
+++ b/arch/gaudi3/arc/onix.c
@@ -3,3 +3,48 @@
  * Copyright (C) 2022 HabanaLabs Ltd.
  * All Rights Reserved.
  */
+
+#include <arcboot.h>
+#include <memory.h>
+
+/*
+ * Because of ONIX RAM size and layout limitation,
+ * both MEM_LKD_PCI and MEM_FWUPD_PCI, are overlapped.
+ *
+ * This table is used by the arcboot_rom_dev_spec_mem_region.
+ * the original function is defined as a weak function
+ *
+ */
+static struct mem_desc mem_desc[MEM_NOENT] = {
+	[MEM_GP] = {
+		.base = MEM_GP_BASE,
+		.size = MEM_GP_SZ,
+	},
+	[MEM_LKD_PCI] = {
+		.base = MEM_LKD_PCI_BASE,
+		.size = MEM_LKD_PCI_SZ,
+	},
+	[MEM_FWUPD_PCI] = {
+		.base = MEM_LKD_PCI_BASE,
+		.size = MEM_LKD_PCI_SZ,
+	},
+	[MEM_I2C] = {
+		.base = MEM_GP_BASE,
+		.size = MEM_GP_SZ,
+	},
+};
+
+int arcboot_rom_dev_spec_mem_region(enum mem_region region,
+					struct mem_desc *desc)
+{
+	if (region >= MEM_NOENT)
+		return 1;
+
+	if (!desc)
+		return 1;
+
+	desc->base = mem_desc[region].base;
+	desc->size = mem_desc[region].size;
+
+	return 0;
+}
diff --git a/common/memory.c b/common/memory.c
index 46a97ec..be008e3 100644
--- a/common/memory.c
+++ b/common/memory.c
@@ -27,7 +27,7 @@
 	},
 };
 
-__weak int arcboot_get_dev_spec_mem_region(__unused enum mem_region region,
+__weak int arcboot_rom_dev_spec_mem_region(__unused enum mem_region region,
 					   __unused struct mem_desc *desc)
 {
 	return 1; /* not dev spec */
@@ -51,7 +51,7 @@
 		return 1;
 
 	/* device defined memory regions */
-	err = arcboot_get_dev_spec_mem_region(region, desc);
+	err = arcboot_rom_dev_spec_mem_region(region, desc);
 	if (!err)
 		return err;
 
